File: 力扣/学习计划/剑指 Offer I/022_股票的最大利润/main.go
package main

// https://leetcode-cn.com/problems/gu-piao-de-zui-da-li-run-lcof/

// 暴力
func maxProfit(prices []int) int {
	max := 0
	for i := len(prices) - 1; i >= 0; i-- {
		for j := 0; j < i; j++ {
			tmp := prices[i] - prices[j]
			if tmp > max {
				max = tmp
			}
		}
	}
	return max
}

File: 力扣/学习计划/剑指 Offer I/022_股票的最大利润/main.java
class Solution {
    /**
     * 动态规划
     */
    public int maxProfit(int[] prices) {
        int len = prices.length;
        if (len < 2) return 0;
        int[] dp = new int[len];
        dp[0] = 0;
        int min = prices[0]; 
        for (int i = 1; i < len; i++) {
            if (prices[i] < min)
                min = prices[i];
            // 状态转移方程
            dp[i] = Math.max(dp[i-1], prices[i] - min);
        }
        return dp[len - 1];
    }
    /**
     * 动态规划优化 - 使用一个变量即可记录结果
     */
    public int maxProfit1(int[] prices) {
        if (prices.length < 2) return 0;
        int min = prices[0], profit = 0;
        for (int i = 0; i < prices.length; i++) {
            if (prices[i-1] < min) 
                min = prices[i-1];
            profit = Math.max(profit, prices[i] - min);
        }
        return profit;
    }
}